Cursor management cleanup

oleg-derevenetz opened the pull request:
Cleanup cursor management (mostly, there are still some places that count on the cursor visibility as a redraw indicator). As a continuation of #3402.

ihhub commented:
Hi @oleg-derevenetz , this is a great effort of cleaning up the code mess. However, I have some notes about the changes:
- **most** of places do not requires cursor hiding / showing logic. It is redundant. The previous implementation was using such logic while rendering a frame (literally it was the case): hide cursor, generate frame, show cursor, render frame and repeat. It was done in such horrible way due to a fact that the old display class had no connection with cursor class. Right now everything is handled by `Display` class
- we should not show cursor for video playback (it is already handled in game_video.cpp) and for popup windows when a user right clicks on something. It should work only on SDL 1 while SDL 2 without software rendering will always show the cursor. In the future we can extend such logic to SDL 2 but right now since SDL 2 cursor is on OS level we can't do much with current implementation
- for the rest of places we have to restore only cursor type so might need 2 classes: one for cursor hiding/showing and another one for cursor type setup and restoring. Also we do not need to pass the cursor as the newly created classes can easily call the singleton inside. This will make code even more cleaner
- for the rest of cases where we need to receive some notification about rendering we have to pass a variable or something. I looked at this code before, it's not trivial for sure with the current implementation.

I hope it'll help you with implementation. Please ask any questions if you aren't sure about something, I'm always open to help :)

oleg-derevenetz commented:
Hi @ihhub

> most of places do not requires cursor hiding / showing logic

Yes, I did my best to remove Hide()/Show() calls from those places like that:

```
cursor.Hide();
blah.Redraw();
cursor.Show();
```

However, it's not so trivial, because in some places cursor state intentionally used as redraw indicator after event processing. I plan to leave such places as is so far and try to do something with them later, after testing existing changes.

> for the rest of places we have to restore only cursor type so might need 2 classes

I use the single class because in most cases "popups" that should hide cursor and "dialogs" with the button are created by the same function. It's handy to use one class to remember both visibility and theme, because sometimes visibility may change and sometimes not, depending on function arguments. With separate restorers it's easier to forgot either of them :)

> Also we do not need to pass the cursor as the newly created classes

Yes, I already thinked about this. It seems that also it's handy to call render() immediately after changing cursor on SDL1, right? There is a slight pause otherwise before software cursor appears or changes its theme.

ihhub commented:
> Hi @ihhub
> 
> > most of places do not requires cursor hiding / showing logic
> 
> Yes, I did my best to remove Hide()/Show() calls from places like that:
> 
> ```
> cursor.Hide();
> blah.Redraw();
> cursor.Show();
> ```
> 
> However, it's not so trivial, because in some places cursor state intentionally used as redraw indicator after event processing. I plan to leave such places as is so far and try to do something with them later, after testing existing changes.
> 
> > for the rest of places we have to restore only cursor type so might need 2 classes
> 
> I use the single class because in most cases "popups" that should hide cursor and "dialogs" with the button are created by the same function. It's handy to use one class to remember both visibility and theme, because sometimes visibility may change and sometimes not, depending on function arguments. With separate restorers it's easier to forgot either of them :)
> 
> > Also we do not need to pass the cursor as the newly created classes
> 
> Yes, I already thought about this. It seems that also it's handy to call render() immediately after changing cursor on SDL1, right? There is a slight pause otherwise before software cursor appears or changes its theme.

What kind of delay are we taking about? We shouldn't redraw the full screen just for cursor change.

oleg-derevenetz commented:
@ihhub

> What kind of delay are we taking about? We shouldn't redraw the full screen just for cursor change.

I mean that with SDL1 cursor, for example, doesn't appear on screen immediately after call of `cursor.Show()`. There is a slight pause about 1/4 sec. However if `render()` is called after `Show()`, it appears on screen immediately. FPS indicator shows ~4.5 FPS on most screens in idle state, it seems that there is some refresh is done (on timer?) that causes cursor to appear (albeit with a delay)?

oleg-derevenetz commented:
Hi @ihhub

> What kind of delay are we taking about? We shouldn't redraw the full screen just for cursor change.

Currently I call Cursor::Redraw() (which effectively calls Display::render()) in Cursor::Show() as a workaround to make cursor appear immediately. However, if there is better solution for this, I would be happy to get some advice.

oleg-derevenetz commented:
After performing a series of tests it seems that `Display::render()` calls are essential for immediate redraw of SDL1 software cursor after changing of theme or visibility (it doesn't hide or appear immediately without this), but `render()` should not be called from `Cursor::Hide()`/`Cursor::Show()` directly, because this causes FPS growth with some visual components like `ListBox` that use those calls for maintaining state after event processing. So I decided to move `render()` call to `~CursorRestorer()`.

ihhub commented on src/fheroes2/gui/cursor.cpp:
```diff
@@ -254,3 +254,34 @@ int Cursor::WithoutDistanceThemes( int theme )
+
+        // immediately render cursor area in case of software emulated cursor
+        if ( fheroes2::cursor().isSoftwareEmulation() ) {
+            const fheroes2::Point pos = LocalEvent::Get().GetMouseCursor();
```
Please use const reference here as `GetMouseCursor` returns such. In this case we can avoit unnessesary object creation.

oleg-derevenetz commented:
OK, what have I got so far:

1. Implemented a new class, `CursorRestorer`, that provides cursor visibility/theme saving using RAII idiom, useful for modal dialogs/popups (and other places, where the cursor is temporarily manipulated, too). Migrated all suitable functions and classes to use this class.
2. Implemented cursor hiding in most pop-up windows (SDL1 only, will not work with SDL2).
3. Removed a lot of redundant `Cursor::Hide()`/`Cursor::Show()` calls.

There are still a few places left where cursor state is used as a redraw indicator, but I think that it's time to review, widely test and merge existing changes. I thoroughly tested these changes myself on MacOS both with SDL1 and SDL2, but I could always miss something. @ihhub @Branikolog I would really appreciate reviewing the existing changes and testing those changes, especially with SDL1 (because hiding the cursor only works with SDL1). Special attention should be paid to the possible cursor visibility issues, FPS spikes and "hide interface" mode.

ihhub commented:
Hi @oleg-derevenetz , I see that we have a lot of places with this code:
```
const CursorRestorer cursorRestorer;
Cursor & cursor = Cursor::Get();
cursor.SetThemes( cursor.POINTER );
cursor.Show();
```

we could definitely reduce code duplication by adding for example a derived from `CursorRestorer` class to call `SetThemes` and `Show` methods within its constructor. Also could we please try to avoid the default value in `CursorRestorer` constructor argument to make things clear. What do you think?

oleg-derevenetz commented:
Hi @ihhub 

> we could definitely reduce code duplication by adding for example a derived from CursorRestorer class to call SetThemes and Show methods within its constructor

Yes, it's an initial setup for a cursor in dialogs. However, it may as well looks like that (for example):

```
const CursorRestorer cursorRestorer;
Cursor & cursor = Cursor::Get();
cursor.SetThemes( cursor.POINTER );
if (buttons) {
    cursor.Show();
}
else {
    cursor.Hide();
}
```

especially in dialogs that also may serve as popups. As an alternative, I can make special constructor for `CursorRestorer` that takes two arguments - `theme` and `show` and apply those after saving of current cursor settings. What do you think?

> Also could we please try to avoid the default value in CursorRestorer constructor argument to make things clear.

There is a special case where `CursorRestorer` is used to fixup mess after `ListBox`, and in this case calling `render()` inside of destructor is a waste of resources, that's why that constructor argument (with a default value) was introduced. I can make a constructor overload instead that takes the same argument without default value. What do you think?

ihhub commented:
> Hi @ihhub
> 
> > we could definitely reduce code duplication by adding for example a derived from CursorRestorer class to call SetThemes and Show methods within its constructor
> 
> Yes, it's an initial setup for a cursor in dialogs. However, it may as well looks like that (for example):
> 
> ```
> const CursorRestorer cursorRestorer;
> Cursor & cursor = Cursor::Get();
> cursor.SetThemes( cursor.POINTER );
> if (buttons) {
>     cursor.Show();
> }
> else {
>     cursor.Hide();
> }
> ```
> 
> especially in dialogs that also may serve as popups. As an alternative, I can make special constructor for `CursorRestorer` that takes two arguments - `theme` and `show` and apply those after saving of current cursor settings. What do you think?
> 
> > Also could we please try to avoid the default value in CursorRestorer constructor argument to make things clear.
> 
> There is a special case where `CursorRestorer` is used to fixup mess after `ListBox`, and in this case calling `render()` inside of destructor is a waste of resources, that's why that constructor argument (with a default value) was introduced. I can make a constructor overload instead that takes the same argument without default value. What do you think?

I think have an extra constructor with `theme` will make code cleaner. About the deafult value of the constructor: it makes sense now, let's keep it as it is.

oleg-derevenetz commented:
@ihhub

> I think having an extra constructor with `theme` will make code cleaner.

Good idea, thanks! Implemented additional overload for convenience.

ihhub commented:
Hi @Branikolog , could you please extensively test this pull request at first priority?

ihhub commented:
Hi @oleg-derevenetz , I tested this pull request. Things which went wrong:
- while switching between heroes the cursor disappears for 1 frame
- switching between towns shows the map for 1 frame
- select a town in the list and then select a hero in the list - cursor is gone

oleg-derevenetz commented:
Hi @ihhub

> while switching between heroes the cursor disappears for 1 frame

By "switching" you mean switching by using hero list on the adventure map screen or by using left/right buttons on the bottom of the hero view screen?

> switching between towns shows the map for 1 frame

The same question. By "shows the map" I suppose that you use eft/right buttons on the bottom of the castle view screen?

> select a town in the list and then select a hero in the list - cursor is gone

Cannot reproduce it so far... But yes, when switching using hero/castle list cursor seems to hide for a short time (but then reappear), looks like another ListBox issue.

ihhub commented:
> Hi @ihhub
> 
> > while switching between heroes the cursor disappears for 1 frame
> 
> By "switching" you mean switching by using hero list on the adventure map screen or by using left/right buttons on the bottom of the hero view screen?
> 
> > switching between towns shows the map for 1 frame
> 
> The same question. By "shows the map" I suppose that you use left/right buttons on the bottom of the castle view screen?
> 
> > select a town in the list and then select a hero in the list - cursor is gone
> 
> Cannot reproduce it so far... But yes, when switching using hero/castle list cursor seems to hide for a short time (but then reappear), looks like another ListBox issue.

My apologies for not being precise.
- I used mouse to click on heroes in the list of heroes on the adventure map
- I used arrows at the bottom of the castle dialog

I'm testing on SDL 1 and indebug mode with slow rendering speed (just set very high resolution).

oleg-derevenetz commented:
Hi @ihhub

> with slow rendering speed (just set very high resolution).

Ah, that's how you notice this... What resolution do you use? I'll going to try the same.

> switching between towns shows the map for 1 frame

Wasn't this observed with master branch? I don't see yet how this may be caused by cursor-related changes...

oleg-derevenetz commented:
@ihhub

> also the hardware on which I tested it is slow: I use Intel embedded GPU

Sometimes it's good to have a slow hardware... :(

oleg-derevenetz commented:
Hi @ihhub, I hope that I fixed these issues.

oleg-derevenetz commented:
@ihhub BTW, what's the point in having "Use dynamic interface for castles" option if it doesn't work and all castle screens (as well as hero screens) are always using "good" interface?

ihhub commented:
> @ihhub BTW, what's the point in having "Use dynamic interface for castles" option if it doesn't work and all castle screens (as well as hero screens) are always using "good" interface?

Haha! You discovered another chunk of code leftovers. Previous team implemented a lot of extra experimental options and not all of them work. We removed like 30 of them but still a lot more left.

ihhub commented:
Hi @oleg-derevenetz , I tested the latest build. 3 things I noticed so far:
- after loading the map and showing the black screen and just before rendering the first frame the cursor disappears
- the same problem for hero's screen as for castle's one: flickering while switching
- cursor still flickers while switching between towns

oleg-derevenetz commented:
Hi @ihhub 

> - after loading the map and showing the black screen and just before rendering the first frame the cursor disappears

I think that's because it is hidden by default. It is hidden on app startup (on the splash screen), then CursorRestorer shows it on all subsequent screens, but "restores" it back when map is loaded, and then it is shown again when adventure map mechanics is started. Is this an issue? 

> - the same problem for hero's screen as for castle's one: flickering while switching

By flickering do you mean flickering of map or cursor? 

> - cursor still flickers while switching between towns

By switching do you mean switching via adventure map or via next/prev buttons on castle view screen?

oleg-derevenetz commented:
Hi @ihhub

> the same problem for hero's screen as for castle's one: flickering while switching

Should be fixed now (if I understand the issue correctly). The same cause as with castle screen. Funny, but I really couldn't notice flickering on my devices.

> cursor still flickers while switching between towns

Should be fixed now as well. That's due to ListBox mechanics - it likes to hide cursor very much.

oleg-derevenetz commented:
Hi @ihhub @Branikolog If there are no more questions about this PR yet, I propose to merge it. Along with #3451 (which can be merged after this one, if there will be no issues with it) it will allow to test cursor mechanics more thoroughly and on a broader scale, including SDL2 builds.

Branikolog commented:
Considering this PR invovles the whole game, I believe we need a long-term testing of this fix... 
Unfortunately, I cannot dedicate much time for testing the nearest few days... So we either need more time or test it after merging... Next release is not so close, so we have some time to fix new bugs. ;)

oleg-derevenetz commented:
@Branikolog

> Next release is not so close, so we have some time to fix new bugs. ;)

That's what I'm talking about too. Together with #3451 all cursor mechanics will be tested more comprehensive and, so to speak, automatically.

ihhub commented:
> @Branikolog
> 
> > Next release is not so close, so we have some time to fix new bugs. ;)
> 
> That's what I'm talking about too. Together with #3451 all cursor mechanics will be tested more comprehensive and, so to speak, automatically during a regular game process.

I'll test these changes myself very soon.

ihhub commented:
Hi @oleg-derevenetz , after short tests: the cursor still blinks while switching towns for example. I think it happens because we hide the cursor first, then we render a new town castle's screen, and after show the cursor back with extra render. Ideally, we should never hide the cursor except for specific windows like pop-up ones.

oleg-derevenetz commented:
Hi @ihhub

> Hi @oleg-derevenetz , after short tests: the cursor still blinks while switching towns for example. I think it happens because we hide the cursor first, then we render a new town castle's screen, and after show the cursor back with extra render. Ideally, we should never hide the cursor except for specific windows like pop-up ones.

Castle screen is one of those screens I didn't touch yet (except inserting CursorRestorer in the very beginning of castle dialog implementation). It have a complex logic with redraw dependent on cursor visibility, so I left it as is so far.

ihhub commented:
> Hi @ihhub
> 
> > Hi @oleg-derevenetz , after short tests: the cursor still blinks while switching towns for example. I think it happens because we hide the cursor first, then we render a new town castle's screen, and after show the cursor back with extra render. Ideally, we should never hide the cursor except for specific windows like pop-up ones.
> 
> Castle screen is one of those screens I didn't touch yet (except inserting CursorRestorer in the very beginning of castle dialog implementation). It have a complex logic with redraw dependent on cursor visibility, so I left it as is so far.
> 
> P.S. By adding debug `printf()`'s you can see that `Castle::OpenDialog()` constantly calls `Cursor::Hide()` and `Cursor::Show()` in its event loop. I plan to try to fix this in future (as well as `ListBox` behavior), but not in this PR. It's already too huge.

I haven't finished my reivew but I think we need to fix everything in one shot :( I'll check it later tomorrow to see what we can do to avoid extra rendering.

ihhub commented:
@oleg-derevenetz , to avoid the problem with cursor hide/show logic without huge changes in the code one of the ideas was to introduce a separate singleton class responsible for this. Of course it'll be a half solution but it could be good for us right now. What do you think?

My big apologies for the delay of this pull request review.

oleg-derevenetz commented:
@ihhub

> one of the ideas was to introduce a separate singleton class responsible for this

What do you mean (may be some example)? Some temporary `RedrawIndicator` with static methods like `ScheduleRedraw()` and `DoneRedraw()` that will be called instead of `Cursor::Hide()` and `CursorShow()`?

ihhub commented:
> @ihhub
> 
> > one of the ideas was to introduce a separate singleton class responsible for this
> 
> What do you mean (may be some example)? Some temporary `RedrawScheduler` with static methods like `ScheduleRedraw()` and `DoneRedraw()` that will be called instead of `Cursor::Hide()` and `Cursor::Show()`?

That's right. We could have a separate file with only 3functions (written below code is for illustration only):
```
namespace fheroes2
{
    namespace globalRedraw
    {
        void reset();
        void schedule();
        bool isScheduled();
    }
}
```
and inside cpp file we'll have a single boolean variable:
```
namespace
{
    bool isGlobalRedrawRequired = false;
}

namespace fheroes2
{
    namespace globalRedraw
    {
        void reset()
        {
            isGlobalRedrawRequired = false;
        }

        void schedule()
        {
            isGlobalRedrawRequired = true;
        }

        bool isScheduled()
        {
            return isGlobalRedrawRequired;
        }
    }
}
```
In this case the code will be easier to read and less confusing. In this case we might even remove
```
// immediately render cursor area in case of software emulated cursor
if ( _renderOnRestore &amp;&amp; fheroes2::cursor().isSoftwareEmulation() ) {
    const fheroes2::Point &amp; pos = LocalEvent::Get().GetMouseCursor(); 
    fheroes2::Display::instance().render( fheroes2::Rect( pos.x, pos.y, 1, 1 ) );
}
```
code to avoid extra rendering.

oleg-derevenetz commented:
@ihhub 

> In this case we might even remove
> 
> ```
> // immediately render cursor area in case of software emulated cursor
> if ( _renderOnRestore &amp;&amp; fheroes2::cursor().isSoftwareEmulation() ) {
>     const fheroes2::Point &amp; pos = LocalEvent::Get().GetMouseCursor(); 
>     fheroes2::Display::instance().render( fheroes2::Rect( pos.x, pos.y, 1, 1 ) );
> }
> ```
> 
> code to avoid extra rendering.

This is not to render screen as a whole, but to render cursor itself after hiding/restoring it if it is software-emulated, so really we can't remove it. Regarding some global singleton/static class - need to think... Looks ugly of course.

oleg-derevenetz commented:
Hi @ihhub OK, I rewrote all places at once where cursor was used as redraw indicator (mostly places where ListBox is used - i.e. file lists, map lists, castle lists and so on). It seems that there is no need for singleton class... However, there may be regressions, because there is a huge pull request with lots of subtle changes in different places.

BTW, cursor is still hidden by default (you may see it on splash screen). This is done intentionally for the reason that I want to catch possible places where the cursor can be still hidden. When all these changes will be well-tested, we can then make a separate PR to show cursor by default right at the application startup (if you need it at all by then).

ihhub commented:
Hi @oleg-derevenetz , I've tested everything and to avoid all my concerns I'm proposing these changes:
1. **cursor.h** file
```
class CursorRestorer
{
public:
    CursorRestorer();
    ~CursorRestorer();

private:
    const int _theme;
};

class CursorSetup
{
public:
    explicit CursorSetup( const int theme );
    ~CursorSetup();

private:
    const int _theme;
    const bool _isVisible;
};
```
2. **cursor.cpp** file:
```
CursorRestorer::CursorRestorer()
    : _theme( Cursor::Get().Themes() )
{
    Cursor & cursor = Cursor::Get();
    cursor.Hide();
}

CursorRestorer::~CursorRestorer()
{
    Cursor::Get().Show();
}

CursorSetup::CursorSetup( const int theme )
    : _theme( Cursor::Get().Themes() )
    , _isVisible( Cursor::Get().isVisible() )
{
    Cursor::Get().SetThemes( theme );
    Cursor::Get().Show();
}

CursorSetup::~CursorSetup()
{
    if ( _isVisible ) {
        Cursor::Get().SetThemes( _theme );
    }
}
```
3. All occurence of `const CursorRestorer cursorRestorer( true, Cursor::POINTER );` replace by `const CursorSetup cursorSetup( Cursor::POINTER );`
4. For situations of presence / absence of buttons in the dialogs:
```
    std::unique_ptr<CursorRestorer> cursorRestorer;
    std::unique_ptr<CursorSetup> cursorSetup;

    if ( showButtons ) {
        cursorSetup.reset( new CursorSetup( Cursor::POINTER ) );
    }
    else {
        cursorRestorer.reset( new CursorRestorer );
    }
```
5. Building info is a special case so at the end of the `bool BuildingInfo::DialogBuyBuilding( bool buttons ) const` method please add this code (the problem with this case is that the caller of it doesn't render after. Alternatively, we could render from the caller side):
```
    if ( !buttons ) {
        cursorRestorer.reset();
        const fheroes2::Point & cursorPos = le.GetMouseCursor();
        Cursor::Get().Redraw( cursorPos.x, cursorPos.y );
    }
```
6. For `Video::ShowVideo` function the changes should be:
```
        std::unique_ptr<CursorRestorer> cursorRestorer;
        if ( roi.empty() ) {
            cursorRestorer.reset( new CursorRestorer );
        }
        else {
            Cursor::Get().setVideoPlaybackCursor();
        }
```
Please let me know if you have any concerns about this solution. My goal is to reduce the number of redrawings as much as possible as they're heavy calls on low-profile machines like PSVita or phones (eventually). I've tested almost every general case and it works fine. We might need to do some adjustments to immediately refresh cursor if there's a delay for such action right now.

oleg-derevenetz commented:
Hi @ihhub

> My goal is to reduce the number of redrawings as much as possible as they're heavy calls on low-profile machines like PSVita or phones (eventually).

Are you sure that this will help much? Every single movement of software cursor means `render()` call anyway (you can turn on FPS counter in game built with SDL1 and move mouse cursor over screen to ensure) and in the case of SDL2 hardware cursor `render()` is never called in CursorRestorer destructor anyway. Also, my goal was to make approach universal regardless of caller, while your approach counts on the fact that cursor will be always shown in the caller, and only cursor theme should be changed.

ihhub commented:
> Hi @ihhub
> 
> > My goal is to reduce the number of redrawings as much as possible as they're heavy calls on low-profile machines like PSVita or phones (eventually).
> 
> Are you sure that this will help much? Every single movement of software cursor means `render()` call anyway (you can turn on FPS counter in game built with SDL1 and move mouse cursor over screen to ensure) and in the case of SDL2 hardware cursor `render()` is never called in CursorRestorer destructor anyway (and AFAIK PSV uses SDL2). Also, my goal was to make approach universal regardless of caller, while your approach counts on the fact that cursor will be always shown in the caller, and only cursor theme should be changed.

Hmm... of course you're right. Let me think a bit more :(

ihhub commented:
@oleg-derevenetz , now I remember why I proposed latest changes: the current CursorRestorer class remembers the last state of the cursor (hide / show) which leads to flickering in some cases (which you observed before). I'm okay to call rendering in destructors if the state of cursor changes (show hidden cursor or the cursor them changed, only these 2 scenarios) but we shouldn't revert to hidden state the cursor. For example, in case of Main Menu once I press a new game button the cursor is again set to hidden state which is incorrect. That's why I separated into 2 classes. If we want to preserve a single class then the current CursorRestorer needs some changes to combine all the logic of my 2 classes.

Does it make sense? :)

oleg-derevenetz commented:
@ihhub

> @oleg-derevenetz , now I remember why I proposed latest changes: the current CursorRestorer class remembers the last state of the cursor (hide / show) which leads to flickering in some cases (which you observed before). I'm okay to call rendering in destructors if the state of cursor changes (show hidden cursor or the cursor them changed, only these 2 scenarios) but we shouldn't revert to hidden state the cursor. For example, in case of Main Menu once I press a new game button the cursor is again set to hidden state which is incorrect. That's why I separated into 2 classes. If we want to preserve a single class then the current CursorRestorer needs some changes to combine all the logic of my 2 classes.
> 
> Does it make sense? :)

This is due to the fact that cursor is hidden by default (just look at its default constructor). I have plan to just call Cursor::Show() somewhere in main() (or just update default constructor - the particular way is debatable) but I need to catch possible places that don't use CursorRestorer first. If wide tests will not find such places, then we will be able to show cursor by default. Currently CursorRestorer destructor have a check that prevents doing something if previous cursor have the same settings as current, so no extra changes are required.

ihhub commented:
> @ihhub
> 
> > @oleg-derevenetz , now I remember why I proposed latest changes: the current CursorRestorer class remembers the last state of the cursor (hide / show) which leads to flickering in some cases (which you observed before). I'm okay to call rendering in destructors if the state of cursor changes (show hidden cursor or the cursor them changed, only these 2 scenarios) but we shouldn't revert to hidden state the cursor. For example, in case of Main Menu once I press a new game button the cursor is again set to hidden state which is incorrect. That's why I separated into 2 classes. If we want to preserve a single class then the current CursorRestorer needs some changes to combine all the logic of my 2 classes.
> > Does it make sense? :)
> 
> This is due to the fact that cursor is hidden by default (just look at its default constructor). I have plan to just call Cursor::Show() somewhere in main() (or just update default constructor - the particular way is debatable) but I need to catch possible places that don't use CursorRestorer first. If wide tests will not find such places, then we will be able to show cursor by default. Currently CursorRestorer destructor have a check that prevents doing something if previous cursor have the same settings as current, so no extra changes are required.

Let's please fix Main menu so we could proceed with these changes because in main menu it's visible that cursor hides before showing for example Credits window.

oleg-derevenetz commented:
@ihhub

> Let's please fix Main menu so we could proceed with these changes because in main menu it's visible that cursor hides before showing for example Credits window.

Than we should make a cursor visible by default so `CursorRestorer` s in all functions that draw main menu windows did nothing in destructors. Should cursor be shown on splash screen and credits screen as well?

ihhub commented:
> @ihhub
> 
> > Let's please fix Main menu so we could proceed with these changes because in main menu it's visible that cursor hides before showing for example Credits window.
> 
> Then we should make a cursor visible by default so `CursorRestorer` s in all functions that draw main menu windows did nothing in destructors. Should cursor be shown on splash screen and credits screen as well?

Please add `Cursor::Get().Show();` line just before `Game::mainGameLoop( isFirstGameRun );` in **fheroes2.cpp** file. We could also add a comment `// Show the cursor at the start on main menu screen`.

ihhub commented:
> Should cursor be shown on splash screen and credits screen as well?
Splash screen and intro video shouln't have cursor.

ihhub commented:
@oleg-derevenetz , let's do even better!
1. Make Show / Hide methods private
2. Make CursorRestorer as a friend class to Cursor class.
3. In `Battle::Interface::Interface` constructor initialize its member as `, _cursorRestorer( true, Cursor::WAR_POINTER )`
4. For game_video.cpp file we need these changes:
```
        std::unique_ptr<CursorRestorer> cursorRestorer;

        if ( roi.empty() ) {
            cursorRestorer.reset( new CursorRestorer( false, Cursor::Get().Themes() ) );
        }
        else {
            cursorRestorer.reset( new CursorRestorer );
            Cursor::Get().setVideoPlaybackCursor();
        }
```
5. In fheroes2.cpp file remove setup of cursor theme (`Cursor::Get().SetThemes( Cursor::POINTER );`) and add this line before `Game::mainGameLoop( isFirstGameRun );`
```
const CursorRestorer cursorRestorer( true, Cursor::POINTER );
```

So we fully prevent any damage to be cause by cursor state and also won't show cursor at the splash screen and intro video for SDL 2 :)

What do you think?

ihhub commented on src/fheroes2/game/game_credits.cpp:
```diff
@@ -34,6 +35,9 @@ void Game::ShowCredits()
     const fheroes2::StandardWindow border( display.DEFAULT_WIDTH, display.DEFAULT_HEIGHT );
 
+    // setup cursor
+    const CursorRestorer cursorRestorer( false, Cursor::Get().Themes() );
```
Please revert these changes for credits. This is misunderstand: the startup has splash screen and intro video while credits is just a normal dialog which can be smaller than the main menu.

oleg-derevenetz replied:
Ah, OK. Credits screen really didn't have cursor before, so I decided to hide it.

ihhub commented:
@oleg-derevenetz , amazing job! Great thanks for doing this!
